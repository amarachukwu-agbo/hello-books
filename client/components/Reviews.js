import React, { Component } from 'react';

class Review extends Component {
  constructor(props) {
    super(props);
    this.renderReviews = this.renderReviews.bind(this);
  }
  renderReviews() {
    const { reviews } = this.props;
    if (!reviews.length) {
      return (
                <div className = "row center">
                   <p className="flow-text"> No Reviews yet.</p>
                </div>
      );
    }
    return reviews.map((review, index) => (
                    <div className="card-panel grey lighten-5" key = { index }>
                            <h6><i className="material-icons prefix">account_circle</i>{ review.userId }
                            <span className="right"> { review.createdAt.split('T')[0] } </span> </h6>
                            <span className="bold">
                                { review.review }
                            </span>
                    </div>
    ));
  }
  render() {
    return (
            <div>
                <h5 className="header center-align book-header">Reviews</h5>
                <div className="card-panel">
                    { this.renderReviews() };
                </div>
            </div>
    );
  }
}

export default Review;

