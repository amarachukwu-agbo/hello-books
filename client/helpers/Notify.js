import swal from 'sweetalert';

/**
 * @description class notifies users using alerts
 *
 * @class Notify
 */
class Notify {
  /**
   * @method notifyError
   * @memberof Notify
   * @static
   * @param {string} error
   *
   * @returns {void}
   */
  static notifyError(error) {
    swal('', error, 'error');
  }

  /**
   * @method notifyInfo
   * @memberof Notify
   * @static
   * @param {string} info
   *
   * @returns {void}
   */
  static notifyInfo(info) {
    swal('', info, 'info');
  }

  /**
   * @method notifySuccess
   * @memberof Notify
   * @static
   * @param {string} message
   *
   * @returns {void}
   */
  static notifySuccess(message) {
    swal('', message, 'success');
  }
}

export default Notify;
