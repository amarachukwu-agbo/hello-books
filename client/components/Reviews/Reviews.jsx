import React, { Component } from 'react';
import propTypes from 'prop-types';
import ReviewForm from './Form/ReviewForm.jsx';
import Notify from '../../helpers/Notify';


/** @description container class for reviews and ReviewForm
 *
 * @class Reviews
 *
 * @extends {React.Component}
 */

class Reviews extends Component {
  /**
   * @constructor create an instance of the component
   *
   * @param {object} props properties for the component
   */
  constructor(props) {
    super(props);
    this.renderReviews = this.renderReviews.bind(this);
    this.submitForm = this.submitForm.bind(this);
  }

  /**
   * @memberof Reviews
   * @method submitForm
   * @description handles submission of review form
   * @param {object} values - form values for submission
   *
   * @returns {void}
   */
  submitForm(values) {
    const { book } = this.props;
    if (values.review && values.review.trim() !== '') {
      this.props.reviewBook(book.id, values);
    } else {
      Notify.notifyError('Review must not be empty');
    }
  }

  /**
   * @memberof Reviews
   * @method renderReviews
   * @description renders the reviews
   *
   * @returns {Node} react node containing reviews
   */
  renderReviews() {
    const { reviews } = this.props;
    if (!reviews.length) {
      return (
        <div className="container center">
          <p>No Reviews yet.</p>
        </div>
      );
    }
    return reviews.map(review => (
      <div className="review-box grey lighten-5" key={review.id}>
        <strong>
          <p className="grey-text text-darken-3">
          <i className="material-icons left">account_circle</i>
          {`${review.userReviews.firstName} ${review.userReviews.lastName}`}
          <span className="right"> {review.createdAt.split('T')[0]} </span>
          </p>
        </strong>
        <span className="bold">
          {review.review}
        </span>
      </div>
    ));
  }

  render() {
    return (
      <div>
        <div className="row">
          <h5 className="header center-align book-header">
            {`Reviews (${this.props.reviews.length})`}
          </h5>
        </div>
        <div className="row z-depth-1">
          {this.props.isAuthenticated &&
            <div className="section">
              <div>
                <ReviewForm submitForm = { this.submitForm }
                  isReviewing = { this.props.isReviewing } />
              </div>
            </div>}
          <div className="review">
            {this.renderReviews()}
          </div>
        </div>
      </div>
    );
  }
}

// Prop type validation
Reviews.propTypes = {
  reviews: propTypes.array,
  isAuthenticated: propTypes.bool,
  isReviewing: propTypes.bool,
  reviewBook: propTypes.func.isRequired,
  book: propTypes.object.isRequired,
};

export default Reviews;

