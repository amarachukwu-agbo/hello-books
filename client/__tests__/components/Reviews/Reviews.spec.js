import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';
import Notify from '../../../helpers/Notify';
import Reviews from '../../../components/Reviews/Reviews';
import mockData from '../../__mocks__/mockData';

let wrapper;
const { books } = mockData;
const NotifyError = sinon.spy(Notify, 'notifyError');

const props = {
  book: books.reviewBookResponse.reviewedBook,
  reviews: books.reviewBookResponse.reviewedBook.bookReviews,
  isAuthenticated: true,
  isReviewing: false,
  reviewBook: sinon.spy(() => Promise.resolve()),
};

sinon.spy(Reviews.prototype, 'submitForm');

describe('<Reviews />', () => {
  afterEach(() => {
    props.reviewBook.reset();
  });

  beforeEach(() => {
    wrapper = shallow(<Reviews {...props} />);
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('calls the submitForm() method', () => {
    const values = {
      review: 'I love this book',
    };
    wrapper.instance().submitForm(values);
    expect(Reviews.prototype.submitForm.called).toEqual(true);
    expect(props.reviewBook.called).toEqual(true);
  });

  it(`calls the submitForm() method and notifies user
    with an error if review is empty`, () => {
    const values = {
      review: '',
    };
    wrapper.instance().submitForm(values);
    expect(Reviews.prototype.submitForm.called).toEqual(true);
    expect(NotifyError.called).toEqual(true);
    expect(NotifyError.calledWith('Review must not be empty'))
      .toEqual(true);
    expect(props.reviewBook.called).toEqual(false);
  });

  it('renders p if there are no reviews', () => {
    wrapper.setProps({
      reviews: [],
      book: books.bookDataResponse.book,
    });
    expect(wrapper.find('p').text()).toEqual('No Reviews yet.');
    expect(wrapper.find('.review-box')).toHaveLength(0);
  });
});
