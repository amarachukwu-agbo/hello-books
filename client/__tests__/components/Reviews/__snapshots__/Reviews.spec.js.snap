// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Reviews /> renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Reviews
    book={
      Object {
        "author": "Joanna Gaines",
        "bookReviews": Array [
          Object {
            "bookId": 1,
            "createdAt": "2018-06-24T13:18:52.295Z",
            "id": 1,
            "review": "I enjoyed reading this book",
            "updatedAt": "2018-06-24T13:18:52.295Z",
            "userId": 3,
            "userReviews": Object {
              "createdAt": "2018-06-23T14:42:46.827Z",
              "firstName": "Grace",
              "id": 3,
              "imageURL": null,
              "lastName": "Jade",
            },
          },
        ],
        "borrowCount": 0,
        "createdAt": "2018-05-10T03:11:52.181Z",
        "description": "A new book",
        "downvotes": 1,
        "favCount": 1,
        "imageURL": "https://images-na.ssl-image.jpg",
        "quantity": 6,
        "subject": "Recipe",
        "title": "A new book",
        "updatedAt": "2018-06-23T19:14:10.925Z",
        "upvotes": 0,
      }
    }
    isAuthenticated={true}
    isReviewing={false}
    reviewBook={[Function]}
    reviews={
      Array [
        Object {
          "bookId": 1,
          "createdAt": "2018-06-24T13:18:52.295Z",
          "id": 1,
          "review": "I enjoyed reading this book",
          "updatedAt": "2018-06-24T13:18:52.295Z",
          "userId": 3,
          "userReviews": Object {
            "createdAt": "2018-06-23T14:42:46.827Z",
            "firstName": "Grace",
            "id": 3,
            "imageURL": null,
            "lastName": "Jade",
          },
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="row"
        >
          <h5
            className="header center-align book-header"
          >
            Reviews (1)
          </h5>
        </div>,
        <div
          className="row z-depth-1"
        >
          <div
            className="section"
          >
            <div>
              <ReduxForm
                isReviewing={false}
                submitForm={[Function]}
              />
            </div>
          </div>
          <div
            className="review"
          >
            <div
              className="review-box grey lighten-5"
            >
              <strong>
                <p
                  className="grey-text text-darken-3"
                >
                  <i
                    className="material-icons left"
                  >
                    account_circle
                  </i>
                  Grace Jade
                  <span
                    className="right"
                  >
                     
                    2018-06-24
                     
                  </span>
                </p>
              </strong>
              <span
                className="bold"
              >
                I enjoyed reading this book
              </span>
            </div>
          </div>
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <h5
            className="header center-align book-header"
          >
            Reviews (1)
          </h5>,
          "className": "row",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Reviews (1)",
            "className": "header center-align book-header",
          },
          "ref": null,
          "rendered": "Reviews (1)",
          "type": "h5",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="section"
            >
              <div>
                <ReduxForm
                  isReviewing={false}
                  submitForm={[Function]}
                />
              </div>
            </div>,
            <div
              className="review"
            >
              <div
                className="review-box grey lighten-5"
              >
                <strong>
                  <p
                    className="grey-text text-darken-3"
                  >
                    <i
                      className="material-icons left"
                    >
                      account_circle
                    </i>
                    Grace Jade
                    <span
                      className="right"
                    >
                       
                      2018-06-24
                       
                    </span>
                  </p>
                </strong>
                <span
                  className="bold"
                >
                  I enjoyed reading this book
                </span>
              </div>
            </div>,
          ],
          "className": "row z-depth-1",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <div>
                <ReduxForm
                  isReviewing={false}
                  submitForm={[Function]}
                />
              </div>,
              "className": "section",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <ReduxForm
                  isReviewing={false}
                  submitForm={[Function]}
                />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "isReviewing": false,
                  "submitForm": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "div",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="review-box grey lighten-5"
                >
                  <strong>
                    <p
                      className="grey-text text-darken-3"
                    >
                      <i
                        className="material-icons left"
                      >
                        account_circle
                      </i>
                      Grace Jade
                      <span
                        className="right"
                      >
                         
                        2018-06-24
                         
                      </span>
                    </p>
                  </strong>
                  <span
                    className="bold"
                  >
                    I enjoyed reading this book
                  </span>
                </div>,
              ],
              "className": "review",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "1",
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <strong>
                      <p
                        className="grey-text text-darken-3"
                      >
                        <i
                          className="material-icons left"
                        >
                          account_circle
                        </i>
                        Grace Jade
                        <span
                          className="right"
                        >
                           
                          2018-06-24
                           
                        </span>
                      </p>
                    </strong>,
                    <span
                      className="bold"
                    >
                      I enjoyed reading this book
                    </span>,
                  ],
                  "className": "review-box grey lighten-5",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": <p
                        className="grey-text text-darken-3"
                      >
                        <i
                          className="material-icons left"
                        >
                          account_circle
                        </i>
                        Grace Jade
                        <span
                          className="right"
                        >
                           
                          2018-06-24
                           
                        </span>
                      </p>,
                    },
                    "ref": null,
                    "rendered": Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": Array [
                          <i
                            className="material-icons left"
                          >
                            account_circle
                          </i>,
                          "Grace Jade",
                          <span
                            className="right"
                          >
                             
                            2018-06-24
                             
                          </span>,
                        ],
                        "className": "grey-text text-darken-3",
                      },
                      "ref": null,
                      "rendered": Array [
                        Object {
                          "instance": null,
                          "key": undefined,
                          "nodeType": "host",
                          "props": Object {
                            "children": "account_circle",
                            "className": "material-icons left",
                          },
                          "ref": null,
                          "rendered": "account_circle",
                          "type": "i",
                        },
                        "Grace Jade",
                        Object {
                          "instance": null,
                          "key": undefined,
                          "nodeType": "host",
                          "props": Object {
                            "children": Array [
                              " ",
                              "2018-06-24",
                              " ",
                            ],
                            "className": "right",
                          },
                          "ref": null,
                          "rendered": Array [
                            " ",
                            "2018-06-24",
                            " ",
                          ],
                          "type": "span",
                        },
                      ],
                      "type": "p",
                    },
                    "type": "strong",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "I enjoyed reading this book",
                      "className": "bold",
                    },
                    "ref": null,
                    "rendered": "I enjoyed reading this book",
                    "type": "span",
                  },
                ],
                "type": "div",
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="row"
          >
            <h5
              className="header center-align book-header"
            >
              Reviews (1)
            </h5>
          </div>,
          <div
            className="row z-depth-1"
          >
            <div
              className="section"
            >
              <div>
                <ReduxForm
                  isReviewing={false}
                  submitForm={[Function]}
                />
              </div>
            </div>
            <div
              className="review"
            >
              <div
                className="review-box grey lighten-5"
              >
                <strong>
                  <p
                    className="grey-text text-darken-3"
                  >
                    <i
                      className="material-icons left"
                    >
                      account_circle
                    </i>
                    Grace Jade
                    <span
                      className="right"
                    >
                       
                      2018-06-24
                       
                    </span>
                  </p>
                </strong>
                <span
                  className="bold"
                >
                  I enjoyed reading this book
                </span>
              </div>
            </div>
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h5
              className="header center-align book-header"
            >
              Reviews (1)
            </h5>,
            "className": "row",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Reviews (1)",
              "className": "header center-align book-header",
            },
            "ref": null,
            "rendered": "Reviews (1)",
            "type": "h5",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="section"
              >
                <div>
                  <ReduxForm
                    isReviewing={false}
                    submitForm={[Function]}
                  />
                </div>
              </div>,
              <div
                className="review"
              >
                <div
                  className="review-box grey lighten-5"
                >
                  <strong>
                    <p
                      className="grey-text text-darken-3"
                    >
                      <i
                        className="material-icons left"
                      >
                        account_circle
                      </i>
                      Grace Jade
                      <span
                        className="right"
                      >
                         
                        2018-06-24
                         
                      </span>
                    </p>
                  </strong>
                  <span
                    className="bold"
                  >
                    I enjoyed reading this book
                  </span>
                </div>
              </div>,
            ],
            "className": "row z-depth-1",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <div>
                  <ReduxForm
                    isReviewing={false}
                    submitForm={[Function]}
                  />
                </div>,
                "className": "section",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <ReduxForm
                    isReviewing={false}
                    submitForm={[Function]}
                  />,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "isReviewing": false,
                    "submitForm": [Function],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": "div",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <div
                    className="review-box grey lighten-5"
                  >
                    <strong>
                      <p
                        className="grey-text text-darken-3"
                      >
                        <i
                          className="material-icons left"
                        >
                          account_circle
                        </i>
                        Grace Jade
                        <span
                          className="right"
                        >
                           
                          2018-06-24
                           
                        </span>
                      </p>
                    </strong>
                    <span
                      className="bold"
                    >
                      I enjoyed reading this book
                    </span>
                  </div>,
                ],
                "className": "review",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "1",
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <strong>
                        <p
                          className="grey-text text-darken-3"
                        >
                          <i
                            className="material-icons left"
                          >
                            account_circle
                          </i>
                          Grace Jade
                          <span
                            className="right"
                          >
                             
                            2018-06-24
                             
                          </span>
                        </p>
                      </strong>,
                      <span
                        className="bold"
                      >
                        I enjoyed reading this book
                      </span>,
                    ],
                    "className": "review-box grey lighten-5",
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": <p
                          className="grey-text text-darken-3"
                        >
                          <i
                            className="material-icons left"
                          >
                            account_circle
                          </i>
                          Grace Jade
                          <span
                            className="right"
                          >
                             
                            2018-06-24
                             
                          </span>
                        </p>,
                      },
                      "ref": null,
                      "rendered": Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "children": Array [
                            <i
                              className="material-icons left"
                            >
                              account_circle
                            </i>,
                            "Grace Jade",
                            <span
                              className="right"
                            >
                               
                              2018-06-24
                               
                            </span>,
                          ],
                          "className": "grey-text text-darken-3",
                        },
                        "ref": null,
                        "rendered": Array [
                          Object {
                            "instance": null,
                            "key": undefined,
                            "nodeType": "host",
                            "props": Object {
                              "children": "account_circle",
                              "className": "material-icons left",
                            },
                            "ref": null,
                            "rendered": "account_circle",
                            "type": "i",
                          },
                          "Grace Jade",
                          Object {
                            "instance": null,
                            "key": undefined,
                            "nodeType": "host",
                            "props": Object {
                              "children": Array [
                                " ",
                                "2018-06-24",
                                " ",
                              ],
                              "className": "right",
                            },
                            "ref": null,
                            "rendered": Array [
                              " ",
                              "2018-06-24",
                              " ",
                            ],
                            "type": "span",
                          },
                        ],
                        "type": "p",
                      },
                      "type": "strong",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "I enjoyed reading this book",
                        "className": "bold",
                      },
                      "ref": null,
                      "rendered": "I enjoyed reading this book",
                      "type": "span",
                    },
                  ],
                  "type": "div",
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
